### DISCUSSION TITLE
MOOSE compilation and unable to use debugger build executable due Lib mesh error

### DISCUSSION THREAD

**Discussion by salaudeen-ya** [idaholab/moose#24724] [Q&A General] [answered]
Hello,

I am trying to build an executable to debug my application. By using `METHOD=dbg make -j1`, I got the following error:

```
Linking Library /home/salau/projects/moose/modules/peridynamics/lib/libperidynamics-dbg.la...
libtool:   error: '/home/salau/projects/moose/modules/peridynamics/build/unity_src/nodalkernels_Unity.x86_64-conda-linux-gnu.dbg.lo' is not a valid libtool object
make: *** [/home/salau/projects/moose/framework/app.mk:345: /home/salau/projects/moose/modules/peridynamics/lib/libperidynamics-dbg.la] Error 1
```

How do I go about this?

**Comment by GiudGiud** [comment]
Hello

Git clean the repo then try to build again

Guillaume

**Reply by salaudeen-ya** [reply]
Thank you, I have been trying this and the build is still not succesful.


In fact, I even try to install a new MOOSE which also fails to build. The error I got is this:

```
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/base/ResidualObject.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/base/MaxVarNDofsPerNode.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/base/Attributes.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/base/Factory.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/base/MooseError.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/utils/MooseEnum.C...
/home/salau/projects/moose/framework/src/base/MooseApp.C: In constructor 'MooseApp::MooseApp(InputParameters)':
/home/salau/projects/moose/framework/src/base/MooseApp.C:411:48: error: 'class libMesh::Parallel::Communicator' has no member named 'sync_type'
  411 |   const_cast<Parallel::Communicator &>(comm()).sync_type(timpi_sync);   
      |                                                ^~~~~~~~~
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/utils/SolutionInvalidityRegistry.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/utils/LeastSquaresFitBase.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/src/utils/MooseEnumBase.C...
make: *** [/home/salau/projects/moose/framework/build.mk:148: /home/salau/projects/moose/framework/src/base/MooseApp.x86_64-conda-linux-gnu.opt.lo] Error 1     
make: *** Waiting for unfinished jobs....
```

**Reply by GiudGiud** [reply]
let s update libmesh and moose again. 
This looks like MOOSE is more up to date than libmesh
see this page
https://mooseframework.inl.gov/getting_started/new_users.html#update

**Reply by salaudeen-ya** [reply]
Okay. I will wait for both the libmesh and moose to be updated.

**Reply by GiudGiud** [reply]
you need to update locally.
We are in sync at all times with the repo. So a brand new install should always work.

The problem arose because you updated libmesh (through conda) or MOOSE (through git)

**Reply by salaudeen-ya** [reply]
I have done that a couple of times today and am still getting a failed build.

I have a brand new install:
```
cd ~/projects
rm -rf moose
git clone https://github.com/idaholab/moose.git
cd moose
git checkout master

```

The installation works fine and I build using:
```
cd ~/projects/moose/test
make -j 4

```

This still give the following error:
```
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/build/unity_src/parser_Unity.C...
Compiling C++ (in opt mode) /home/salau/projects/moose/framework/build/unity_src/relationshipmanagers_Unity.C...
x86_64-conda-linux-gnu-c++: fatal error: Killed signal terminated program cc1plus
compilation terminated.
make: *** [/home/salau/projects/moose/framework/build.mk:148: /home/salau/projects/moose/framework/build/unity_src/systems_Unity.x86_64-conda-linux-gnu.opt.lo] Error 1
make: *** Waiting for unfinished jobs....
```

**Reply by GiudGiud** [reply, marked as answer]
this is yet another error message though

x86_64-conda-linux-gnu-c++: fatal error: Killed signal terminated program cc1plus

usually means you are running out of memory. Try make -j 2 and if not, make -j 1

**Reply by salaudeen-ya** [reply]
Hello,

I am trying to use the traditional method of debugging the libmesh by setting up several print statements in the code using standard ostream. For now, I am modifying the fe_map.C.  I followed up with the procedures of updating the libmesh by running the script `update_and_rebuild_libmesh.sh ` as found here:

https://mooseframework.inl.gov/getting_started/installation/build_libmesh.html
https://mooseframework.inl.gov/help/faq/libmesh_issues.html

Is there any other thing to do to make the changes effective? Since it neither catches the deliberate error I introduced in the print statement nor has the line printed on the terminal when I test a simple 1D input file.

**Reply by GiudGiud** [reply]
are you on mac or linux?
and when building libmesh, did you `skip-submodule-update` ?

**Reply by salaudeen-ya** [reply]
Yes, 
I did try `(base) salau@LAPTOP-9A1UOG6G:~/projects/moose/scripts$ ./update_and_rebuild_libmesh.sh --skip-submodule-update`.

I am on Windows Subsystem for Linux (WSL).

**Reply by GiudGiud** [reply]
If you `ldd moose_test-opt` or whatever executable you are building, do you see the expected libmesh in moose/libmesh

Are you sure that the code you are writing in libmesh is being executed?

**Reply by salaudeen-ya** [reply]
By using `ldd moose_test-opt`:

```
(moose) salau@LAPTOP-9A1UOG6G:~/projects/ddm$ ldd ddm-opt
        linux-vdso.so.1 (0x00007f8f0829c000)
        libddm_test-opt.so.0 => /home/salau/projects/ddm/test/lib/libddm_test-opt.so.0 (0x00007f8f08280000)
        libddm-opt.so.0 => /home/salau/projects/ddm/lib/libddm-opt.so.0 (0x00007f8f08275000)
        libmodule_loader_with-opt.so.0 => /home/salau/projects/moose/modules/module_loader/lib/libmodule_loader_with-opt.so.0 (0x00007f8f08266000)
        libmoose-opt.so.0 => /home/salau/projects/moose/framework/libmoose-opt.so.0 (0x00007f8f060f9000)
        libpcre-opt.so.0 => /home/salau/projects/moose/framework/contrib/pcre/libpcre-opt.so.0 (0x00007f8f060cd000)
        libhit-opt.so.0 => /home/salau/projects/moose/framework/contrib/hit/libhit-opt.so.0 (0x00007f8f0606f000)
        libmesh_opt.so.0 => /home/salau/miniconda3/envs/moose/libmesh/lib/libmesh_opt.so.0 (0x00007f8f04a81000)
        libtimpi_opt.so.11 => /home/salau/miniconda3/envs/moose/libmesh/lib/libtimpi_opt.so.11 (0x00007f8f04a72000)
        libhdf5_cpp.so.200 => /home/salau/miniconda3/envs/moose/lib/libhdf5_cpp.so.200 (0x00007f8f049fa000)
        libvtkIOCore-9.2.so.1 => /home/salau/miniconda3/envs/moose/libmesh-vtk/lib/libvtkIOCore-9.2.so.1 (0x00007f8f0498a000)
...
```
Basically stating the effective libmesh header is in the `conda` environment.

The code I have been writing in the libmesh is not executed and I have been modifying the libmesh in `/projects/moose/libmesh`

**Reply by GiudGiud** [reply]
That s your problem then
DIsable the conda environment, clean MOOSE, then try to install again

**Reply by salaudeen-ya** [reply]
Hello, 

I am still stuck on this and the following provides a detailed explanation of what I have:

•	I cleaned my Linux by removing both the MOOSE and my conda environment
•	Then, I proceeded to do a clean installation of both the environment and the MOOSE
•	After cloning and building MOOSE, my local “libmesh” library is empty
•	Yet, my application (input files) works perfectly even when I used quadrature rules which I know is implemented in the libmesh.
•	This makes me wonder if the Libmesh referenced by MOOSE is hosted on online repository and only accessed with the header file
•	It is noteworthy that this is a recent development because I had a locally modified libmesh before when I was implementing some specific quadrature rules.
•	Although the local “libmesh” library is empty, there is an option for running a script that will download (update and build) the libmesh framework to the local repository. After doing this, any changes I make to the libmesh library won't be captured still.


What am I doing wrong?

**Reply by GiudGiud** [reply]
> • After cloning and building MOOSE, my local “libmesh” library is empty


libmesh is not moose/libmesh if you use mamba.
It s in a mamba package

So basically do not use mamba's moose-libmesh / moose-dev package if you plan to modify libmesh.

Instead, use the update_and_rebuild_libmesh.sh script in moose/scripts

**Reply by salaudeen-ya** [reply]
I actually used conda installation and not mamba.
These are my results:

```(moose) salau@LAPTOP-9A1UOG6G:~/projects/moose$ ./scripts/update_and_rebuild_libmesh.sh
ERROR: LIBMESH_DIR=/home/salau/miniconda3/envs/moose/libmesh exists within the conda base /home/salau/miniconda3
Exiting in order to stop you from overwriting your conda environment
(moose) salau@LAPTOP-9A1UOG6G:~/projects/moose$ cd libmesh/
(moose) salau@LAPTOP-9A1UOG6G:~/projects/moose/libmesh$ ls

...

(moose) salau@LAPTOP-9A1UOG6G:~/miniconda3/envs/moose/libmesh$ ls
Make.common  bin  contrib  etc  examples  include  lib  share
(moose) salau@LAPTOP-9A1UOG6G:~/miniconda3/envs/moose/libmesh$ cd include/
(moose) salau@LAPTOP-9A1UOG6G:~/miniconda3/envs/moose/libmesh/include$ ls
Eigen  gzstream.h  metaphysicl  netcdf_aux.h  netcdf_meta.h  poly2tri  unsupported
boost  libmesh     netcdf.h     netcdf_mem.h  netcdf_par.h   timpi
```

Trying to update and rebuild the libmesh throws an error, the libmesh library is empty on my local repository, and this is what I could get from the libmesh reference in the conda environment.

**Reply by GiudGiud** [reply]
conda and mamba end up doing the same thing

everything works as expected. You CANNOT use a custom libmesh if you are using the libmesh from conda. 

You need to follow the instructions 
https://mooseframework.inl.gov/help/faq/faq_build_libmesh.html

**Reply by salaudeen-ya** [reply]
I followed the instruction on compiling libmesh by running the `update_and_rebuild_libmesh.sh`. I tried modifying the local libmesh library and deliberately introducing errors in some files. By running the `update_and_rebuild_libmesh.sh`, it only checks for things like the mpi, OpenCL, HDF5, zlib, compiler and makes some optional configurations. 

I expected the build to fail due to the deliberate errors I introduced in some of the libmesh files (for me to be sure it's compiling correctly) but it does not. Do you know how I can solve that?

**Reply by GiudGiud** [reply]
The problem is that the script checks out a new version of libmesh when you run it 

you need to run the script with ‘—skip-submodule-update ´ on the command line
